diff --git a/src/protocol.c b/src/protocol.c
index 53e65d4..86a0268 100644
--- a/src/protocol.c
+++ b/src/protocol.c
@@ -315,8 +315,13 @@ int callback_tty(struct lws *wsi, enum lws_callback_reasons reason, void *user,
           break;
         case RESIZE_TERMINAL:
           if (pss->process == NULL) break;
-          json_object_put(
-              parse_window_size(pss->buffer + 1, pss->len - 1, &pss->process->columns, &pss->process->rows));
+          {
+            json_object *_o = parse_window_size(pss->buffer + 1, pss->len - 1, &pss->process->columns, &pss->process->rows);
+            struct json_object *_p = NULL;
+            if (json_object_object_get_ex(_o, "xpixel", &_p)) pss->process->xpixel = (uint16_t)json_object_get_int(_p);
+            if (json_object_object_get_ex(_o, "ypixel", &_p)) pss->process->ypixel = (uint16_t)json_object_get_int(_p);
+            json_object_put(_o);
+          }
           pty_resize(pss->process);
           break;
         case PAUSE:
diff --git a/src/pty.c b/src/pty.c
index 01e9d6b..f1521cf 100644
--- a/src/pty.c
+++ b/src/pty.c
@@ -150,7 +150,7 @@ bool pty_resize(pty_process *process) {
   COORD size = {(int16_t) process->columns, (int16_t) process->rows};
   return pResizePseudoConsole(process->pty, size) == S_OK;
 #else
-  struct winsize size = {process->rows, process->columns, 0, 0};
+  struct winsize size = {process->rows, process->columns, process->xpixel, process->ypixel};
   return ioctl(process->pty, TIOCSWINSZ, &size) == 0;
 #endif
 }
diff --git a/src/pty.h b/src/pty.h
index c3f761c..265903b 100644
--- a/src/pty.h
+++ b/src/pty.h
@@ -29,6 +29,7 @@ typedef void (*pty_exit_cb)(pty_process *);
 struct pty_process_ {
   int pid, exit_code, exit_signal;
   uint16_t columns, rows;
+  uint16_t xpixel, ypixel;
 #ifdef _WIN32
   STARTUPINFOEXW si;
   HPCON pty;
